doc/sphinx/proof-engine/ltac.rst

```diff
@@ -844,66 +961,126 @@ We can carry out pattern matching on terms with:
            end.
+
+.. _match_vs_lazymatch_ex:
+
+   .. example:: Comparison of match, multimatch and lazymatch
+
+      **todo: how do match and multimatch differ?**
+
+      .. coqtop:: reset in
+
          Goal True.
-         f (3+4).
+
+      .. coqtop:: all
+
+         match False with
+         | False => idtac "try"; fail
+         | _ => easy
+         end.
+
+      .. coqtop:: reset in
+
+         Goal True.
+
+      .. coqtop:: all
+
+         multimatch False with
+         | False => idtac "try"; fail
+         | _ => easy
+         end.
+
+      .. coqtop:: reset in
+
+         Goal True.
+
+      .. coqtop:: all fail
+
+         lazymatch False with
+         | False => idtac "try"; fail
+         | _ => easy
+         end.
 
 .. _ltac-match-goal:
 
-Pattern matching on goals
-~~~~~~~~~~~~~~~~~~~~~~~~~
+Pattern matching on goals: match goal
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-We can perform pattern matching on goals using the following expression:
+.. todo PR The title seems incomplete, should be "goals and hypotheses", right?
 
-.. we should provide the full grammar here
+Use this form to match against goals.  These patterns have zero or
+more subpatterns to match hypotheses followed by a subpattern to match
+the conclusion.
 
-.. tacn:: match goal with {+| {+, @context_hyp} |- @cpattern => @ltac_expr } | _ => @ltac_expr end
-   :name: match goal
+.. tacn:: @match_key {? reverse } goal with {? %| } {+| @match_context_rule } end
+   :name: match goal; lazymatch goal; multimatch goal
 
-   If each hypothesis pattern :n:`hyp`\ :sub:`1,i`, with i = 1, ..., m\ :sub:`1` is
+   .. insertprodn match_context_rule match_hyp
+
+   .. prodn::
+      match_context_rule ::= {*, @match_hyp } |- @match_pattern => @ltac_expr
+      | [ {*, @match_hyp } |- @match_pattern ] => @ltac_expr
+      | _ => @ltac_expr
+      match_hyp ::= @name : @match_pattern
+      | @name := {? [ @match_pattern ] : } @match_pattern
+
+   :tacn:`match goal`, :tacn:`lazymatch goal` and :tacn:`multimatch goal` are :token:`ltac_expr1`\s.
+
+   `multimatch` (in :token:`match_key`)
+      Allows subsequent tactics
+      to backtrack into a right-hand side tactic which has backtracking points
+      left and trigger the selection of a new matching branch or combination of
+      hypotheses when all the backtracking points of the right-hand side have
+      been consumed.
+
+      :n:`match { reverse } goal …` is, in fact, shorthand for
+      :n:`once multimatch { reverse } goal …`.
+
+   `lazymatch` (in :token:`match_key`)
+      Causes the match to commit to the first matching branch with the first
+      matching combination of hypotheses rather than trying a new match if
+      the right-hand side fails. If the right-hand side of the selected branch
+      is a tactic with backtracking points, then subsequent failures cause
+      this tactic to backtrack.
+
+   If each hypothesis pattern :n:`match_hyp`\ :sub:`1,i`, with i = 1, ..., m\ :sub:`1` is
```

Comment:
```suggestion
   If each hypothesis pattern :n:`@match_hyp`\ :sub:`1,i`, with i = 1, ..., m\ :sub:`1` is
```

Reply:
Fixed.  An unrelated question:  I didn't expect to see "Tactic failure" here.  What's going on?

```
(multimatch True with
| True => idtac "branch 1"
| _ => idtac "branch 2"
end ;
match True with
| True => idtac "branch A"; fail
end) ; idtac.

Output:
Tactic failure.
branch 1
branch A
branch 2
branch A

```

Reply:
I can only imagine that your output comes from an improper interleaving of stdout and stderr.  The actual output is
```
branch 1
branch A
branch 2
branch A
File "/tmp/foo.v", line 2, characters 2-161:
Error: Tactic failure.
```
The `Tactic failure` is a result of your `fail`, and I don't see why you expect things to succeed overall.

> It makes me wonder whether the backtracking mechanism behaves consistently enough for us to document.

Backtracking is pretty consistent, and definitely worth documenting.  The only confusing part is the interaction of backtracking with failure level, which indeed I'm not sure if we should document because I don't have a sensible model of it in my head. @ppedrot @aspiwack, can one of you describe how failure levels are intended to interact with backtracking points introduced by `+`?  Consider the following:
```coq
Goal True.
  Fail match goal with | [ |- True ] => idtac end; fail 2. (* Tactic failure (level 2). *)
  Fail lazymatch goal with | [ |- True ] => idtac end; fail 2. (* Tactic failure (level 2). *)
  Fail multimatch goal with | [ |- True ] => idtac end; fail 2. (* Tactic failure (level 1). *)
  Fail idtac; fail 2. (* Tactic failure (level 2). *)
  Fail (idtac + idtac); fail 2. (* Tactic failure (level 1). *)
  Fail (fail 2 + idtac). (* Tactic failure (level 1). *)
  Fail (fail 2 + idtac) + idtac. (* Tactic failure. *)
  Fail fail 2 + (idtac + idtac). (* Tactic failure (level 1). *)
  Fail fail 2 + idtac + idtac. (* Tactic failure (level 1). *)
```
As I understand it, the behavior is something like "when backtracking across a `fail`, reduce the failure level by 1; if the failure level is now 0, try the next option in the stream of successes; if the failure level is still non-zero, then discard the remaining successes and fail"

@jfehrle The purpose of failure level is to break out of nested backtracking: for example, `first [ a; fail 1 | b ]` will not try `b` if `a` succeeds and leaves over any goals (the entire tactic will instead fail), while it will try `b` when `a` fails.